File: java/rna-transcription/src/main/java/RnaTranscription.java
import java.util.*;

class RnaTranscription {

    private static final Map<Character, Character> nucleotideComplements;
    static {
        Map<Character, Character> map = new HashMap<>();
        map.put('C', 'G');
        map.put('G', 'C');
        map.put('T', 'A');
        map.put('A', 'U');
        nucleotideComplements = Collections.unmodifiableMap(map);
    }

    String transcribe(String dnaStrand) {
        String rnaTranscription = "";

        char[] nucleotides = dnaStrand.toCharArray();
        for (char nucleotide : nucleotides) {
            rnaTranscription += nucleotideComplements.get(nucleotide);
        }

        return rnaTranscription;
    }

}

File: elixir/rna-transcription/lib/rna_transcription.ex
defmodule RnaTranscription do
  @nucleotide_complements %{
    ?G => ?C,
    ?C => ?G,
    ?T => ?A,
    ?A => ?U
  }

  def fetch_nucleotide_complement(nucleotide) do
    Map.fetch!(@nucleotide_complements, nucleotide)
  end

  @doc """
  Transcribes a character list representing DNA nucleotides to RNA

  ## Examples

  iex> RnaTranscription.to_rna('ACTG')
  'UGAC'
  """
  @spec to_rna([char]) :: [char]
  def to_rna(dna) do
    dna
    |> Enum.map(&fetch_nucleotide_complement/1)
  end
end
